### PR TITLE
MOTECH-2907: Add visit UUID to create encounter task action

### PR DESCRIPTION
**Pull request by palbecki**

### CODE REVIEW

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by pgesek** [comment]
test this please

File: openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java [hunk 1]
```diff
@@ -38,16 +39,17 @@ public Encounter() {
     }
 
     public Encounter(Location location, EncounterType encounterType, Date encounterDatetime, Patient patient,
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java:41]: Please add additional constructor to have backward compatibility

File: openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java [hunk 2]
```diff
@@ -116,27 +126,29 @@ public void setObs(List<Observation> obs) {
...
+                Objects.equals(encounterProviders, encounter.encounterProviders);
+    }
 
-        Encounter other = (Encounter) obj;
-
-        return Objects.equals(this.uuid, other.uuid) && Objects.equals(this.display, other.display) &&
-                Objects.equals(this.location, other.location) && Objects.equals(this.encounterType, other.encounterType) &&
-                Objects.equals(this.encounterDatetime, other.encounterDatetime) && Objects.equals(this.patient, other.patient) &&
-                Objects.equals(this.encounterProviders, other.encounterProviders) && Objects.equals(this.obs, other.obs);
+    @Override
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java]: why do you move hashCode method below equals ?

File: openmrs/src/main/java/org/motechproject/openmrs/domain/Visit.java [hunk 3]
```diff
@@ -15,6 +15,12 @@
     private Date stopDatetime;
     private Patient patient;
 
+    /**
+     * Default constructor.
+     */
+    public Visit() {
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/domain/Visit.java:21]: do we need default constructor ?

File: openmrs/src/main/java/org/motechproject/openmrs/service/impl/OpenMRSVisitServiceImpl.java [hunk 4]
```diff
@@ -38,6 +39,16 @@ public Visit createVisit(String configName, Visit visit) {
         return createdVisit;
     }
 
+    @Override
+    public Visit getVisitByUuid(String configName, String uuid) {
+        try {
+            Config config = configService.getConfigByName(configName);
+            return visitResource.getVisitById(config, uuid);
+        } catch (HttpClientErrorException e) {
+            return null;
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/service/impl/OpenMRSVisitServiceImpl.java:76]: Log error here, similar to other services

File: openmrs/src/main/java/org/motechproject/openmrs/tasks/OpenMRSActionProxyService.java [hunk 5]
```diff
@@ -30,8 +30,9 @@
      * @param providerUuid      the provider uuid
      * @param observations      the map of observations where concept name is a key and value from the observation is a value
      */
-    Encounter createEncounter(String configName, DateTime encounterDatetime, String encounterType, String locationName,
-                              String patientUuid, String providerUuid, Map<String, String> observations);
+    Encounter createEncounter(String configName, DateTime encounterDatetime, String encounterType,
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/tasks/OpenMRSActionProxyService.java:34]: adjust javadoc

File: openmrs/src/main/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceImpl.java [hunk 6]
```diff
@@ -67,18 +67,19 @@
 
     @Override
     public Encounter createEncounter(String configName, DateTime encounterDatetime, String encounterType,
-                                String locationName, String patientUuid, String providerUuid,
+                                String locationName, String patientUuid, String providerUuid, String visitUuid,
                                 Map<String, String> observations) {
         Location location = getLocationByName(configName, locationName);
         Patient patient = patientService.getPatientByUuid(configName, patientUuid);
         Provider provider = providerService.getProviderByUuid(configName, providerUuid);
+        Visit visit = visitService.getVisitByUuid(configName, visitUuid);
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceImpl.java]: visitUuid is optional field so you should handle potential null

File: openmrs/src/test/java/org/motechproject/openmrs/it/MRSTasksIntegrationBundleIT.java [hunk 7]
```diff
@@ -347,7 +347,7 @@ private void createEncounterTestData() throws ParseException, ConceptNameAlready
 
         Location location = locationService.getLocations(DEFAULT_CONFIG_NAME, DEFAULT_LOCATION_NAME).get(0);
         Encounter encounter = new Encounter(location, createdEncounterType, new DateTime("2001-01-16T00:00:00Z").toDate(),
-                createdPatient, Collections.singletonList(createdProvider.getPerson()), null);
+                createdPatient, null, Collections.singletonList(createdProvider.getPerson()), null);
```
Comment (kmadej) [inline openmrs/src/test/java/org/motechproject/openmrs/it/MRSTasksIntegrationBundleIT.java]: include creating visit in ITs

File: openmrs/src/main/java/org/motechproject/openmrs/domain/Visit.java (hunk 3 above)
Comment (palbecki) [reply openmrs/src/main/java/org/motechproject/openmrs/domain/Visit.java:21]: I use it in unit tests.

File: openmrs/src/test/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceTest.java [hunk 8]
```diff
@@ -123,25 +123,31 @@ public void shouldCreateEncounterWithGivenParameters() {
...
         doReturn(patient).when(patientService).getPatientByUuid(eq(CONFIG_NAME), eq(patient.getUuid()));
         doReturn(provider).when(providerService).getProviderByUuid(eq(CONFIG_NAME), eq(provider.getUuid()));
-        doReturn(encounter).when(encounterService).createEncounter(eq(CONFIG_NAME), eq(encounter));
+        doReturn(visit).when(visitService).getVisitByUuid(eq(CONFIG_NAME), eq(visit.getUuid()));
+
         doReturn(Collections.singletonList(location))
                 .when(locationService).getLocations(eq(CONFIG_NAME), eq(location.getName()));
 
-        Encounter encounterCreated = openMRSActionProxyService.createEncounter(CONFIG_NAME, new DateTime(encounter.getEncounterDatetime()),
+        openMRSActionProxyService.createEncounter(CONFIG_NAME, new DateTime(encounter.getEncounterDatetime()),
```
Comment (kmadej) [inline openmrs/src/test/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceTest.java]: do not change this test, it should still check if proper encounter is returned from createEncounter service

File: openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java (hunk 1 above)
Comment (palbecki) [reply openmrs/src/main/java/org/motechproject/openmrs/domain/Encounter.java:41]: done

File: openmrs/src/main/java/org/motechproject/openmrs/tasks/OpenMRSActionProxyService.java (hunk 5 above)
Comment (palbecki) [reply openmrs/src/main/java/org/motechproject/openmrs/tasks/OpenMRSActionProxyService.java:34]: done.

**Review by palbecki — COMMENTED** [review]
fix

File: openmrs/src/test/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceTest.java [hunk 9]
```diff
@@ -123,25 +123,33 @@ public void shouldCreateEncounterWithGivenParameters() {
...
         doReturn(Collections.singletonList(location))
                 .when(locationService).getLocations(eq(CONFIG_NAME), eq(location.getName()));
 
         Encounter encounterCreated = openMRSActionProxyService.createEncounter(CONFIG_NAME, new DateTime(encounter.getEncounterDatetime()),
                 encounter.getEncounterType().getName(), location.getName(), patient.getUuid(), provider.getUuid(),
-                observations);
-        
+                visit.getUuid(), observations);
+
+        verify(encounterService).createEncounter(eq(CONFIG_NAME), encounterCaptor.capture());
```
Comment (kmadej) [inline openmrs/src/test/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceTest.java]: this is redundant

File: openmrs/src/test/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceTest.java [hunk 10]
```diff
@@ -123,25 +123,33 @@ public void shouldCreateEncounterWithGivenParameters() {
...
+
         assertEquals(encounter, encounterCreated);
+        assertEquals(encounter, encounterCaptor.getValue());
```
Comment (kmadej) [inline openmrs/src/test/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceTest.java]: this is redundant

File: openmrs/src/main/java/org/motechproject/openmrs/helper/EventHelper.java [hunk 11]
```diff
@@ -139,6 +140,20 @@ private EventHelper() {
...
+     * @param uuid the visit ID to be parsed
+     * @return the map with single ID parameter
+     */
+    public static Map<String, Object> visitParameters(String uuid) {
+        Map<String, Object> visitParameters = new HashMap<>();
+        visitParameters.put(EventKeys.VISIT_ID, uuid);
+        return visitParameters;
+    }
+
+
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/helper/EventHelper.java]: remove additional line

**Review by kmadej — CHANGES_REQUESTED** [review]

**Review by kmadej — APPROVED** [review]

**Pull request merged**